### ISSUE TITLE
Move pre-start scripts from init containers into the regular container

### ISSUE THREAD

**Issue by jandubois** [cloudfoundry-incubator/cf-operator#890]
By running the pre-start scripts in init containers, they are executed sequentially. In bosh they are all running in parallel.

**Describe the solution you'd like**

We already have the `container-run` mechanism to execute post-start scripts, so we could just extend that tool to add support for pre-start scripts as well.

As a side effect this will also cut down significantly on the total number of init containers, as right now one is created for each job in the instance group. Most of them will simply do nothing if the corresponding job doesn't have a pre-start script.

**Describe alternatives you've considered**

The alternative would be to do nothing. The current mechanism works, and is probably only slightly slower (and has additional clutter in the pod spec for all the additional init containers).

**Comment by manno** [comment]
The container-run command is missing a README, but apparently we use it to run  `a command and a post-start with optional conditions`.
It's also listening to a socket, to start/stop the command?

Now we want to also execute multiple processes in parallel for pre-start?
This sounds more and more like an init system.

**Comment by jandubois** [comment]
> This sounds more and more like an init system.

But isn't that already what it is? 😄 

I do understand the concern, and wouldn't have suggested doing this if we didn't already have container-run for running post-start processes.

If we don't want to do this, then at least the operator should not generate additional init containers for jobs that don't even have a pre-start script. The generated pod specs look more like un-optimized generated machine code than hand-crafted description that are accessible for human inspection.

I agree though that this is rather low priority.

**Comment by manno** [comment]
:)

Note to self:
> the operator should not generate additional init containers for jobs that don't even have a pre-start script

This is about BOSH pre-start https://github.com/cloudfoundry-incubator/cf-operator/blob/master/pkg/bosh/bpmconverter/container_factory.go#L127-L135

**Comment by viovanov** [comment]
Based on measuring init container start times, parallelization of pre-start scripts (both BPM and BOSH kinds) wouldn't help. We've optimized the longest running init containers, but for kubecf long startup times are caused by serial startup.
